cvprac/cvp_api.py

```diff
@@ -3204,6 +3210,449 @@ def workspace_build_status(self, workspace_id, build_id):
+                                     ' from 2021.2.0 or newer.')
+                return None
+        self.log.debug('v6 {}'.format(url))
+        return self.clnt.get(url, timeout=self.request_timeout)
+
+
+    def change_control_get_all(self):
+        ''' Get the configuration and status of all Change Controls using Resource APIs.
+            Supported versions: CVP 2021.2.0 or newer and CVaaS.
+
+            Returns:
+               response (dict): A dict that contains a list of all Change Controls.
+        '''
+        url = '/api/resources/changecontrol/v1/ChangeControl/all'
+        # For on-prem check the version as it is only supported from 2021.2.0+
+        if not self.clnt.is_cvaas:
+            if self.clnt.apiversion is None:
+                self.get_cvp_info()
+            if self.clnt.apiversion < 6.0:
+                self.log.warning('change_control_get_all: Change Control Resource APIs are supported'
+                                     ' from 2021.2.0 or newer.')
+                return None
+        self.log.debug('v6 {}'.format(url))
         return self.clnt.get(url, timeout=self.request_timeout)
+
+
+    def change_control_approval_get_one(self, cc_id, cc_time=None):
+        ''' Get the state of a specific Change Control's approve config using Resource APIs.
+            Supported versions: CVP 2021.2.0 or newer and CVaaS.
+
+            Args:
+               cc_id (str): The ID of the change control.
+               cc_time (str): Time indicates the time for which you are interested in the data.
+                    If no time is given, the server will use the time at which it makes the request.
+                    The time format is RFC 3339, e.g.: 2021-12-24T11:30:00.00Z.
+            Returns:
+               response (dict): A dict that contains...
+                    Ex: {'value': {'key': {'id': '<CC ID>'}, 'approve': {'value': True},
+                         'version': '2021-12-13T21:05:58.813750128Z'}, 'time': '2021-12-13T21:11:26.788753264Z'}
+        '''
+        if cc_time is None:
+            params = 'key.id={}'.format(cc_id)
+        else:
+            params = 'key.id={}&time={}'.format(cc_id, cc_time)
+        url = '/api/resources/changecontrol/v1/ApproveConfig?' + params
+        # For on-prem check the version as it is only supported from 2021.2.0+
+        if not self.clnt.is_cvaas:
+            if self.clnt.apiversion is None:
+                self.get_cvp_info()
+            if self.clnt.apiversion < 6.0:
+                self.log.warning('change_control_approval_get_one: Change Control Resource APIs are supported'
+                                    ' from 2021.2.0 or newer.')
+                return None
+        try:
+            cc_status = self.change_control_get_one(cc_id)
+        except Exception as e:
+            self.log.warning('change_control_get_one: The change control ID does not exist.')
+            return None
+        if 'approve' not in cc_status['value']:
```

Comment:
Interesting case here for if we should check for 'value' key as well or risk an exception.

Ex:

if 'value' not in cc_status or 'approve' not in cc_status['value']:
   ....

Reply:
I think we shoould check for value in cc_status as a mentioned.